CloudFormation updates

wjordan opened the pull request:
This PR updates the CloudFormation templates, preprocessor and other related scripts to be more stable and feature-complete, as a prerequisite to a production-capable auto-scaling setup:
##### `cloud_formation.rb`

Several updates to the preprocessor:
- Allow template and stack names to be passed via environment variables (so `rake adhoc:start STACK_NAME=x TEMPLATE=xyz.yml.erb` can be used to deploy specific stacks from specific templates).
- Support special `%{"Key": "Value"}` syntax for inserting [Intrinsic Functions](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) into processed file contents (via the `file` erb-method).
- Add `lambda` helper function, to simplify the use of [AWS Lambda-backed custom resources](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources-lambda.html).
##### `stack-xref.js`, `lambda.yml.erb`, `iam.yml.erb`, `vpc.yml.erb`

This Lambda-backed custom resource implements the 'service-oriented stack' layered-architecture pattern described in the [Refer to Resources in Another Stack](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/walkthrough-custom-resources-lambda-cross-stack-ref.html) doc:

> Cross-stack references enable you to use a layered or service-oriented architecture by creating related AWS resources in individual stacks instead of including all resources in a single stack, and then referring to needed resources from the appropriate stack.

Using this pattern, I've split out and deployed `lambda`, `IAM` and `VPC` CloudFormation stacks containing Lambda functions for custom resources, IAM access roles/policies and VPC-related networking resources respectively.
##### `ami-manager.js`, `ami_bootstrap.sh.erb`

This Lambda-backed custom resource and associated bootstrap script implements an automatic AMI-creation pattern described in Tom Maddox's AWS DevOps blog post ["Faster Auto Scaling in AWS CloudFormation Stacks with Lambda-backed Custom Resources"](https://blogs.aws.amazon.com/application-management/post/Tx38Z5CAM5WWRXW). This feature allowed for faster auto-scaling by reducing the total time to “scale out” (launch new instances) in response to failed instances or increased load metrics (from ~20 to ~10 minutes), at the cost of an additional layer of complexity (and the additional time needed to create/update the base image, currently ~30 minutes, though a new base image should only be needed very infrequently).

(An alternative for AMI-creation I also explored is to use [Packer](https://www.packer.io/) to create and deploy AMIs through a separate configuration, but the current approach using a Lambda function invoking the EC2 `CreateImage` API directly seemed easier to manage as long as we're dealing with an AWS-only infrastructure.)
##### cloud_formation_stack.yml.erb

Slight updates to the auto-scaling stack template to make use of the cross-referenced stack layers (lambda, IAM and VPC), and to integrate the AMI-creation resources into the stack.

philbogle commented on aws/cloudformation/ami-manager.js:
```diff
@@ -0,0 +1,112 @@
+var response = require('cfn-response');
```
Could you please add a file comment at the head of the file explaining what this module does?  (I think maybe you could just move up the method comment below since this module contains only that one method.)

wjordan replied:
done (used the `@file` JSDoc tag we've started using elsewhere in our JS codebase)

philbogle commented on aws/cloudformation/ami-manager.js:
```diff
@@ -0,0 +1,112 @@
+  var instanceId = event.ResourceProperties.InstanceId;
+  var instanceRegion = event.ResourceProperties.Region;
+
+  var responseStatus = "FAILED";
```
Does the responseStatus need to be hoisted here and mutated?

It only every seems to be used as an argument to response.send (with "FAILED" in one case and "SUCCESS in the other), so it seems like it would be clearer to use the literal strings in the call (or assign a constant right before the call if it's important the parameter have a meaningful name.

wjordan replied:
The cfn-response module actually provides explicitly-enumerated strings for the response status, so I will use those directly

philbogle commented on aws/cloudformation/ami-manager.js:
```diff
@@ -0,0 +1,112 @@
+  var ec2 = new AWS.EC2({region: instanceRegion});
+
+  console.log("REQUEST TYPE:", event.RequestType);
+  if (event.RequestType == "Delete") {
```
It seems the only event.RequestType we handle is Delete., are the others not important?  Could you explicitly check for the other types that we know are safe and thow if there is an unexpected even type?

wjordan replied:
done

philbogle commented on aws/cloudformation/ami-manager.js:
```diff
@@ -0,0 +1,118 @@
+      } else if (data.Images.length === 0) {
+        response.send(event, context, "SUCCESS", {Info: "Nothing to delete"});
+      } else {
+        var imageId = data.Images[0].ImageId;
```
This seems to assume that length is exactly 1.  Is it correct to ignore the other images?

wjordan replied:
We're calling `describeImages` with a specific image ID as a parameter, so if more than one `Image` is returned it should be an error. I'll add a condition for this case.

philbogle commented:
Done with my initial pass, just a few comments.

wjordan commented:
Thanks for the review Phil. I've addressed all of the comments, and added a minify filter to the JavaScript code to keep the `ami-manager.js` under 4096-character limit with the changes.

philbogle commented:
LGTM
